oap-ape/ape-native/src/reader.cc

```diff
@@ -166,20 +166,31 @@ int Reader::readBatch(int32_t batchSize, int64_t* buffersPtr_, int64_t* nullsPtr
   allocateExtraBuffers(batchSize, buffersPtr, nullsPtr, buffersPtr_, nullsPtr_);
 
   currentBatchSize = batchSize;
+  int rowsRet = 0;
+  if (aggExprs.size() == 0) {  // will not do agg
+    int rowsToRead = doReadBatch(batchSize, buffersPtr, nullsPtr);
+    totalRowsRead += rowsToRead;
+    ARROW_LOG(DEBUG) << "total rows read yet: " << totalRowsRead;
+    rowsRet = doFilter(rowsToRead, buffersPtr, nullsPtr);
+  } else {
+    results.resize(aggExprs.size());
+    while (totalRowsRead < totalRows && !checkEndOfRowGroup()) {
+      int rowsToRead = doReadBatch(batchSize, buffersPtr, nullsPtr);
+      totalRowsRead += rowsToRead;
+      ARROW_LOG(INFO) << "total rows read yet: " << totalRowsRead;
 
-  int rowsToRead = doReadBatch(batchSize, buffersPtr, nullsPtr);
-  totalRowsRead += rowsToRead;
-  ARROW_LOG(DEBUG) << "total rows read yet: " << totalRowsRead;
+      int rowsAfterFilter = doFilter(rowsToRead, buffersPtr, nullsPtr);
 
-  int rowsAfterFilter = doFilter(rowsToRead, buffersPtr, nullsPtr);
+      rowsRet = doAggregation(rowsAfterFilter, map, keys, results, buffersPtr, nullsPtr);
+    }
 
-  std::vector<DecimalVector> results;
-  std::vector<Key> keys;
-  ApeHashMap map;
-  int rowsRet = doAggregation(rowsAfterFilter, map, keys, results, buffersPtr, nullsPtr);
-  if (aggExprs.size()) {
-    dumpBufferAfterAgg(groupByExprs.size(), aggExprs.size(), keys, results, buffersPtr_,
-                       nullsPtr_);
+    if (aggExprs.size()) {
+      dumpBufferAfterAgg(groupByExprs.size(), aggExprs.size(), keys, results, buffersPtr_,
```

Comment:
There may be buffer overflows since keys.size() could be larger than batchSize.

Reply:
Thanks for your comments. I didn't consider this case, I will address this issue in another PR.